### PR TITLE
Add cosmetic option to show seed info

### PR DESCRIPTION
**Pull request by rlbond86**
This PR adds a cosmetic option to display the seed information on the file screen. This can be useful for streaming, multiworld, or just to identify the generated ROM.

The information displays the world number (multiworld only), randomizer version, generation time, a user-defined message (up to 48 characters), and displays whether a spoiler was generated or if plandomizer was used. The world number is always shown, even if the cosmetic option is disabled.

Example:

Glide64_THE_LEGEND_OF_ZELDA_22

Glide64_THE_LEGEND_OF_ZELDA_23

### CODE REVIEW

**Comment by ETR-BTF** [comment]
I'm not sure how useful it'd be to display the settings string: it's too long to compare to someone else's or copy into the generator, and the string alone doesn't tell you anything about the settings. Displaying the player ID and world count is really nice tho. I wonder if it'd be nice to put it on the file select screen so you can confirm the player ID/count and hash simultaneously.

**Comment by rlbond86** [comment]
There are several streams that list the settings string though. Although it's of limited usefulness I still think it can occasionally be helpful

**Comment by matthewkirby** [comment]
I love this for the random settings league. Super easy to confirm that the right version of the randomizer was used. Also solves the problem that I am always paranoid about: player number confusion in multiworld haha. 

I expect there might be risks (maybe not though?) but would adding an extra line of user supplied text below the "Generated at..." line be doable (with a 20char limit or something)? Maybe some field that is only usable through plando or something (I dunno if thats frowned upon)? I would love to be able to plando an extra line about the random settings script version used.

**Comment by rlbond86** [comment]
I've added a 32-character custom message option. The settings string is now truncated to 64 characters.

**Comment by rlbond86** [comment]
I've shifted some fields around to make a little more room for seeds.

**Comment by ETR-BTF** [comment]
Is it possible to retrieve the settings used and/or the spoiler log of a race seed if the settings and seed strings are given?

**Comment by fenhl** [comment]
Yes. There should be an option to hide those while still displaying the info at the top.

**Comment by Roman971** [comment]
Providing the generation time *may* be an issue if it helps narrowing down the RNG over the course of a seed. I have no idea if it would be relevant with OoTR and Python's RNG methods, but I heard some randomizers had this kind of issue in the past, where people were able to figure out the rest of the seed based on the approximate generation time and the first set of locations.
Something to keep in mind for races at least.

**Comment by rlbond86** [comment]
> 
> 
> Is it possible to retrieve the settings used and/or the spoiler log of a race seed if the settings and seed strings are given?



> 
> 
> Yes. There should be an option to hide those while still displaying the info at the top.

Isn't that taken care of by selecting "generate race seed"?

**Comment by rlbond86** [comment]
> 
> 
> Providing the generation time _may_ be an issue if it helps narrowing down the RNG over the course of a seed. I have no idea if it would be relevant with OoTR and Python's RNG methods, but I heard some randomizers had this kind of issue in the past, where people were able to figure out the rest of the seed based on the approximate generation time and the first set of locations.
> It's probably not a concern with this rando, but something to keep in mind for races at least.

Maybe only showing HH:MM and randomly adding +/- 1 minute would fix it?

**Comment by ETR-BTF** [comment]
> > Is it possible to retrieve the settings used and/or the spoiler log of a race seed if the settings and seed strings are given?
> 
> > Yes. There should be an option to hide those while still displaying the info at the top.
> 
> Isn't that taken care of by selecting "generate race seed"?

Race seeds can only be generated on the website. It'd be nice if this info could also be hidden on offline versions of the randomizer, both for "casual" races and tourneys that may want to use a version that's not currently on the website (e.g. Roman's branch). Since the hash can already be used to confirm everyone is playing the same seed, which is also easier to compare with, you could just hide this info on _every_ seed. That would also avoid another toggle on the GUI, since the version number and player count (if applicable) is good info to always have turned on imo.

**Comment by rlbond86** [comment]
I've updated the PR with the following:

1. Info is shown on file screen instead of title screen
2. Seed and settings are no longer shown
3. Multiworld info is always shown
4. User-defined string is lengthened to fit ootrandomizer.com URL

**Comment by dragonbane0** [comment]
Looks like Cosmetics.py imposes a max length of 52 onto this setting, but the max_length limit of the GUI restricts the user to 48 characters still, unless this is intentional? 

Also, while I haven't looked over it, but I assume the python code properly enforces the string length and won't lead to unpredictable behaviour no matter the input? Assume the user bypasses the GUI and inputs a 2000 character string or tries weird insertions to mask the string as a broken URL (this is important for web security when it comes to any arbitrary user input hence I have to check).

**Comment by rlbond86** [comment]
The max length of 52 in Cosmetics.py vs 48 in the GUI is intentional -- it allows a little "slack" for line wrapping. This allows `https://ootrandomizer.com/` to fit on a single line.

The string length is strictly enforced. The variable `custom_msg` is limited to the first 52 characters of the setting variable on line 39 of Cosmetics.py.

**Comment by rlbond86** [comment]
I've further clamped CLI input to 48 characters instead of 52.

**Comment by dragonbane0** [comment]
Alright, then this looks good to me!

**Comment by fenhl** [comment]
I checked out this branch and while the seed info shows up correctly, I cannot get the custom message to appear at all. Am I missing something or is this feature unfinished?

File: Cosmetics.py [hunk 1]
```diff
@@ -28,6 +28,33 @@ def patch_dpad(rom, settings, log, symbols):
...
+        msg = [custom_msg[:line_len], custom_msg[line_len:]]
+        part1 = msg[0].split(' ')
+        if len(part1[-1]) + len(msg[1]) < line_len:
+            msg = [" ".join(part1[:-1]), part1[-1] + msg[1]]
+        else:
+        # Is it a URL?
+            part1 = msg[0].split('/')
+            if len(part1[-1]) + len(msg[1]) < line_len:
+                msg = ["/".join(part1[:-1]) + "/", part1[-1] + msg[1]]
+        for idx,part in enumerate(msg):
```
Comment (Zannick) [inline Cosmetics.py:53]: This loop appears to be indented too far, so messages under the line_len don't get used.
Edit: I can't figure out how to make a suggestion on github when I've already commented.

**Review by Zannick — CHANGES_REQUESTED** [review]
1. We should have an early validation of the `user_message` that it is ASCII-only; anything I tested in range 128-255 is silently replaced with a space, and above that causes the patching process to fail (which is wayyy too late to fail based on user input). I was going to suggest using `str.encode` instead of `ord` but encode turns higher characters into multi-byte sequences... and presumably N64 doesn't have all of utf-8 anyway.
2. My particular RetroArch settings disable super-awesome upscaling in favor of performance, and the text is not super-legible. Actually, I played around with the parallel graphics settings that I thought had to do with resolution, but I must have been misremembering which graphics plugin I had used, as none of them seemed to change the resolution of this text meaningfully. Is this a result of the new scaling of the text characters in `text_flush_size`?

Screenshot 2021-02-08 103655

File: Cosmetics.py (hunk 1 above)
Comment (fenhl) [reply Cosmetics.py:53]: Unindenting this loop fixes my issue (My custom message was simply `test`).

**Comment by fenhl** [comment]
An issue with this is that since this is a cosmetic option, patching with “Override Original Cosmetics” also overrides the custom message. This makes the feature a lot less useful for the random settings league. Is there any reason this isn't just a regular non-cosmetic option?

**Comment by Crain-32** [comment]
In the case of Player ID, couldn't one of the top items be one with an associated amount (Bow, Bombs, Rupees?) and then the displayed amount would be your Seed's Player ID.
Multiworld's could then share the same Item Hash, with this one factor changing.

**Comment by cjohnson57** [comment]
As far as I can tell, this PR is superseded by #1253, so I'm going ahead and closing this one- If I'm wrong feel free to reopen.

**Pull request closed without merging**